nixos/modules/system/boot/loader/refind/refind.nix

```diff
@@ -0,0 +1,77 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+
+  cfg = config.boot.loader.refind;
+
+  efi = config.boot.loader.efi;
+
+  refindBuilder = pkgs.substituteAll {
+    src = ./refind-builder.py;
+
+    isExecutable = true;
+
+    inherit (pkgs) python3;
+
+    nix = config.nix.package.out;
+
+    timeout = if config.boot.loader.timeout != null then config.boot.loader.timeout else "";
+
+    extraConfig = cfg.extraConfig;
+
+    extras = if cfg.extras != null then cfg.extras else "";
+
+    inherit (pkgs) refind efibootmgr coreutils gnugrep gnused gawk utillinux;
+
+    inherit (efi) efiSysMountPoint canTouchEfiVariables;
+  };
+
+in {
+
+  options.boot.loader.refind = {
+    enable = mkOption {
+      description = "Whether to enable the refind EFI boot manager";
+      type = types.bool;
+      default = false;
+    };
+
+    extraConfig = mkOption {
+      description = "Extra configuration text appended to refind.conf";
+      type = types.lines;
+      default = "";
+    };
+
+    extras = mkOption {
+      description = "Extra content to be copied to extras-directory.";
+      type = types.nullOr types.path;
+      default = null;
+    };
+  };
+
+  config = mkIf cfg.enable {
+    assertions = [
+      {
+        assertion = (config.boot.kernelPackages.kernel.features or { efiBootStub = true; }) ? efiBootStub;
```

Comment:
This assertion only checks *whether* `features` has an attribute `efiBootStub`, but doesn't assert its value being false. You probably want `! config.boot.kernelPackages.kernel.features.efiBootStub or true` instead, which means the `efiBootStub` feature should be off, with a default value of it being true (so the assertion fails if it's unset)

Reply:
@Infinisil i don't quite understand.
so you are suggesting  `assertion = config.boot.kernelPackages.kernel.features.efiBootStub or true;` instead ?

Reply:
Oh, you seem to have taken this assertion from https://github.com/NixOS/nixpkgs/blob/ea97fe2ce4ef94215a2c51dd220e0c7dd35e6b43/nixos/modules/system/boot/loader/systemd-boot/systemd-boot.nix#L94

But that assertion is wrong, because
```nix
{ efiBootStub = false; } ? efiBootStub == true
```

Because the `?` operator only checks *whether* the left hand side has or doesn't have an attribute, it doesn't return its value.

Compare it to this, which does return the value:
```nix
{ efiBootStub = false; }.efiBootStub == false
```

So for this particular case, you need
```nix
{
  assertion = config.boot.kernelPackages.kernel.features.efiBootStub or true;
  message = "...";
}
```

Which means "The `efiBootStub` value must be true, but if it doesn't exist that's fine too" (the default is `true`, see https://github.com/NixOS/nixpkgs/blob/ea97fe2ce4ef94215a2c51dd220e0c7dd35e6b43/pkgs/os-specific/linux/kernel/generic.nix#L66)

Reply:
done.
as discussed with @Infinisil on IRC: also added a commit to correct the assertion in `systemd-boot`